disnake/ui/action_row.py

```diff
@@ -24,86 +24,169 @@
+
     .. versionadded:: 2.4
 
+    .. versionchanged:: 2.6
+        Requires and provides stricter typing for contained components.
+
     Parameters
     ----------
     *components: :class:`WrappedComponent`
         The components of this action row.
+
+        .. versionchanged:: 2.6
+            Components can now be either valid in the context of a message, or in the
+            context of a modal. Combining components from both contexts is not supported.
     """
 
-    def __init__(self, *components: WrappedComponent):
-        self.width: int = 0
-        raw_components: List[NestedComponent] = []
-        # Validate the components
-        for component in components:
-            if not isinstance(component, WrappedComponent):
-                raise ValueError("ActionRow must contain only WrappedComponent instances")
+    type: ClassVar[Literal[ComponentType.action_row]] = ComponentType.action_row
 
-            self.width += component.width
+    # When unspecified and called empty, default to an ActionRow that takes any kind of component.
 
-            if self.width > 5:
-                raise ValueError("Too many components in one row.")
+    @overload
+    def __init__(self: ActionRow[WrappedComponent]):
+        ...
 
-            raw_components.append(component._underlying)
+    # Explicit definitions are needed to make
+    # "ActionRow(Select(), TextInput())" and
+    # "ActionRow(Select(), Button())"
+    # differentiate themselves properly.
 
-        self._underlying = ActionRowComponent._raw_construct(
-            type=ComponentType.action_row,
-            children=raw_components,
-        )
+    @overload
+    def __init__(self: ActionRow[MessageUIComponent], *components: MessageUIComponent):
+        ...
+
+    @overload
+    def __init__(self: ActionRow[ModalUIComponent], *components: ModalUIComponent):
+        ...
+
+    # Allow use of "ActionRow[StrictUIComponent]" externally.
+
+    @overload
+    def __init__(self: ActionRow[StrictUIComponentT], *components: StrictUIComponentT):
+        ...
+
+    def __init__(self, *components: UIComponentT):  # type: ignore
+        self._children: List[UIComponentT] = []
+
+        for component in components:
+            self.append_item(component)
 
     def __repr__(self) -> str:
         return f"<ActionRow children={self.children!r}>"
 
-    @property
-    def children(self) -> List[NestedComponent]:
-        """List[Union[:class:`~disnake.Button`, :class:`~disnake.SelectMenu`, :class:`~disnake.TextInput`]]: The components of this row."""
-        return self._underlying.children
+    def __len__(self) -> int:
+        return len(self.children)
 
-    @children.setter
-    def children(self, value: List[NestedComponent]):
-        self._underlying.children = value
+    @property
+    def children(self) -> List[UIComponentT]:
+        """List[:class:`WrappedComponent`]: The UI components stored in this action row."""
+        return self._children
```

Comment:
This was originally kept for some semblance of protection of the underlying children. But, as discussed on discord, it has now been replaced with a simple attribute because of how ineffective this was.